# Hi everybody. All three parts of the assignment are in this notebook together. They start with markdown cells "Part 1", "Part 2", "Part 3". Thanks.

# Part 1

In [1]:
# import necessary libraries, working with BeautifulSoup
from bs4 import BeautifulSoup
import requests
import pandas as pd

In [2]:
# download data from wikipedia
List_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(List_url).text
soup = BeautifulSoup(source, 'xml')
table=soup.find('table')

In [3]:
# create new dataframe with the three required columns
column_names = ['Postalcode','Borough','Neighborhood']
df = pd.DataFrame(columns = column_names)

In [4]:
# search through table
for tr_cell in table.find_all('tr'):
    row_data=[]
    for td_cell in tr_cell.find_all('td'):
        row_data.append(td_cell.text.strip())
    if len(row_data)==3:
        df.loc[len(df)] = row_data

In [5]:
df.head()

Postalcode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

Data cleaning.

In [6]:
# remove lines without boroughs
df=df[df['Borough']!='Not assigned']

In [7]:
# set neighborhood as borough if neighborhood not defined
for i in range(0,df.shape[0]):
    if df.iloc[i][2] == 'Not assigned':
        df.iloc[i][2] = df.iloc[i][1]
        i = i+1
#df.head()
df

Postalcode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M6A        North York   
6          M7A  Downtown Toronto   
8          M9A         Etobicoke   
9          M1B       Scarborough   
11         M3B        North York   
12         M4B         East York   
13         M5B  Downtown Toronto   
14         M6B        North York   
17         M9B         Etobicoke   
18         M1C       Scarborough   
20         M3C        North York   
21         M4C         East York   
22         M5C  Downtown Toronto   
23         M6C              York   
26         M9C         Etobicoke   
27         M1E       Scarborough   
30         M4E      East Toronto   
31         M5E  Downtown Toronto   
32         M6E              York   
36         M1G       Scarborough   
39         M4G         East York   
40         M5G  Downtown Toronto   
41         M6G  Downtown Toronto   
45         M1H       Scarborough   
46         M2H        North York   
47         M3H        North York   
48         M4H         East York   
..         ...               ...   
111        M4R   Central Toronto   
112        M5R   Central Toronto   
113        M6R      West Toronto   
114        M7R       Mississauga   
116        M9R         Etobicoke   
117        M1S       Scarborough   
120        M4S   Central Toronto   
121        M5S  Downtown Toronto   
122        M6S      West Toronto   
126        M1T       Scarborough   
129        M4T   Central Toronto   
130        M5T  Downtown Toronto   
135        M1V       Scarborough   
138        M4V   Central Toronto   
139        M5V  Downtown Toronto   
142        M8V         Etobicoke   
143        M9V         Etobicoke   
144        M1W       Scarborough   
147        M4W  Downtown Toronto   
148        M5W  Downtown Toronto   
151        M8W         Etobicoke   
152        M9W         Etobicoke   
153        M1X       Scarborough   
156        M4X  Downtown Toronto   
157        M5X  Downtown Toronto   
160        M8X         Etobicoke   
165        M4Y  Downtown Toronto   
168        M7Y      East Toronto   
169        M8Y         Etobicoke   
178        M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
8              Islington Avenue, Humber Valley Village  
9                                       Malvern, Rouge  
11                                           Don Mills  
12                     Parkview Hill, Woodbine Gardens  
13                            Garden District, Ryerson  
14                                           Glencairn  
17   West Deane Park, Princess Gardens, Martin Grov...  
18              Rouge Hill, Port Union, Highland Creek  
20                                           Don Mills  
21                                    Woodbine Heights  
22                                      St. James Town  
23                                  Humewood-Cedarvale  
26   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                   Guildwood, Morningside, West Hill  
30                                         The Beaches  
31                                         Berczy Park  
32                                 Caledonia-Fairbanks  
36                                              Woburn  
39                                             Leaside  
40                                  Central Bay Street  
41                                            Christie  
45                                           Cedarbrae  
46                                   Hillcrest Village  
47     Bathurst Manor, Wilson Heights, Downsview North  
48                                    Thorncliffe Park  
..  

Remove duplicates.

In [8]:
df2=df.drop_duplicates(subset="Postalcode", keep="first", inplace=False)
df2

Postalcode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M6A        North York   
6          M7A  Downtown Toronto   
8          M9A         Etobicoke   
9          M1B       Scarborough   
11         M3B        North York   
12         M4B         East York   
13         M5B  Downtown Toronto   
14         M6B        North York   
17         M9B         Etobicoke   
18         M1C       Scarborough   
20         M3C        North York   
21         M4C         East York   
22         M5C  Downtown Toronto   
23         M6C              York   
26         M9C         Etobicoke   
27         M1E       Scarborough   
30         M4E      East Toronto   
31         M5E  Downtown Toronto   
32         M6E              York   
36         M1G       Scarborough   
39         M4G         East York   
40         M5G  Downtown Toronto   
41         M6G  Downtown Toronto   
45         M1H       Scarborough   
46         M2H        North York   
47         M3H        North York   
48         M4H         East York   
..         ...               ...   
111        M4R   Central Toronto   
112        M5R   Central Toronto   
113        M6R      West Toronto   
114        M7R       Mississauga   
116        M9R         Etobicoke   
117        M1S       Scarborough   
120        M4S   Central Toronto   
121        M5S  Downtown Toronto   
122        M6S      West Toronto   
126        M1T       Scarborough   
129        M4T   Central Toronto   
130        M5T  Downtown Toronto   
135        M1V       Scarborough   
138        M4V   Central Toronto   
139        M5V  Downtown Toronto   
142        M8V         Etobicoke   
143        M9V         Etobicoke   
144        M1W       Scarborough   
147        M4W  Downtown Toronto   
148        M5W  Downtown Toronto   
151        M8W         Etobicoke   
152        M9W         Etobicoke   
153        M1X       Scarborough   
156        M4X  Downtown Toronto   
157        M5X  Downtown Toronto   
160        M8X         Etobicoke   
165        M4Y  Downtown Toronto   
168        M7Y      East Toronto   
169        M8Y         Etobicoke   
178        M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
8              Islington Avenue, Humber Valley Village  
9                                       Malvern, Rouge  
11                                           Don Mills  
12                     Parkview Hill, Woodbine Gardens  
13                            Garden District, Ryerson  
14                                           Glencairn  
17   West Deane Park, Princess Gardens, Martin Grov...  
18              Rouge Hill, Port Union, Highland Creek  
20                                           Don Mills  
21                                    Woodbine Heights  
22                                      St. James Town  
23                                  Humewood-Cedarvale  
26   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                   Guildwood, Morningside, West Hill  
30                                         The Beaches  
31                                         Berczy Park  
32                                 Caledonia-Fairbanks  
36                                              Woburn  
39                                             Leaside  
40                                  Central Bay Street  
41                                            Christie  
45                                           Cedarbrae  
46                                   Hillcrest Village  
47     Bathurst Manor, Wilson Heights, Downsview North  
48                                    Thorncliffe Park  
..  

In [9]:
df2.shape

(103, 3)

The list has 103 rows, which means there are 103 different postcodes.

# Part 2

In [10]:
geodata_df = pd.read_csv("http://cocl.us/Geospatial_data")

In [11]:
geodata_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [12]:
geodata_df.rename(columns={'Postal Code':'Postalcode'},inplace=True)
geodata_merge = pd.merge(geodata_df, df2, on='Postalcode')

In [13]:
full_table=geodata_merge[['Postalcode','Borough','Neighborhood','Latitude','Longitude']]
full_table.head()

Postalcode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [14]:
full_table.shape

(103, 5)

# Part 3

In [ ]:
import os
from sklearn.cluster import KMeans
!conda install -c conda-forge folium
!conda install -c conda-forge folium=0.11.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
from geopy.geocoders import Nominatim 
import matplotlib.cm as cm
import matplotlib.colors as colors
import numpy as np
print('Libraries have been imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    libgomp-9.3.0              |      h5dbcf3e_17         378 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    certifi-2020.12.5          |   py36h5fab9bb_0         143 KB  conda-forge
    ca-certificates-2020.12.5  |       ha878542_0         137 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge
    libgcc-ng-9.3.0            |      h5dbcf3e_17         7.8 MB  conda-forge
    openssl-1.1.1i             |       h7f98852_0         2.1 MB  conda-forge
    _openmp_mutex-4.5          |            1_gnu          22 KB  conda-forge
    folium-0.11.

In [16]:
toronto_data=full_table[full_table['Borough'].str.contains("Toronto")]
toronto_data.head()

Postalcode          Borough                    Neighborhood   Latitude  \
37        M4E     East Toronto                     The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
43        M4M     East Toronto                 Studio District  43.659526   
44        M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

In [33]:
CLIENT_ID = 'UT2CSTWIIPGBQHESMVORMMZF3Q21K0PKIVOYR10JAKO33TQN' 
CLIENT_SECRET = 'I2ZWM3FHFXTB2O1MRVXRRM2VXNKHPHO2WALYG2MIF1G5HS5J'
VERSION = '20180604'

In [18]:
def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [19]:
toro_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High Park, The Junction South
Parkdale, Ron

In [20]:
toro_venues.head()

Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

In [34]:
# count number of venues per neighborhood
toro_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            54   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre                                  17   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   77   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             38   
Davisville North                                                        7   
Dufferin, Dovercourt Village                                           17   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              56   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               43   
Moore Park, Summerhill East                                             3   
North Toronto West                                                     22   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            36   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                               92   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     38   
St. James Town                                                         77   
St. James Town, Cabbagetown                                            45   
Stn A PO Boxes                                                         93   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    24   
The Beaches                                                             4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             54   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre                                   17   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    77   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              38   
Davisville North                    

In [22]:
# one hot encoding
toro_onehot = pd.get_dummies(toro_venues[['Venue Category']], prefix="", prefix_sep="")
toro_onehot.drop(['Neighborhood'],axis=1,inplace=True) 
toro_onehot.insert(loc=0, column='Neighborhood', value=toro_venues['Neighborhood'] )
toro_onehot.shape

(1613, 239)

In [23]:
toro_grouped = toro_onehot.groupby('Neighborhood').mean().reset_index()
toro_grouped.head()

Neighborhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2              Business reply mail Processing Centre                0.0   
3  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
4                                 Central Bay Street                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0  0.000000            0.000000      0.000000        0.000000   
1  0.000000            0.000000      0.000000        0.000000   
2  0.000000            0.000000      0.000000        0.000000   
3  0.066667            0.066667      0.066667        0.066667   
4  0.000000            0.000000      0.000000        0.000000   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0              0.0          0.000000                  0.0           0.0  ...   
1              0.0          0.000000                  0.0           0.0  ...   
2              0.0          0.000000                  0.0           0.0  ...   
3              0.2          0.133333                  0.0           0.0  ...   
4              0.0          0.000000                  0.0           0.0  ...   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0               0.0    0.0            0.0                       0.018519   
1               0.0    0.0            0.0                       0.000000   
2               0.0    0.0            0.0                       0.000000   
3               0.0    0.0            0.0                       0.000000   
4               0.0    0.0            0.0                       0.015385   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wings Joint  \
0               0.0                    0.0       0.0          0.0   
1               0.0                    0.0       0.0          0.0   
2               0.0                    0.0       0.0          0.0   
3               0.0                    0.0       0.0          0.0   
4               0.0                    0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0     0.000000  
1            0.0     0.000000  
2            0.0     0.000000  
3            0.0     0.000000  
4            0.0     0.015385  

[5 rows x 239 columns]

In [24]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [25]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toro_grouped['Neighborhood']

for ind in np.arange(toro_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toro_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2              Business reply mail Processing Centre           Pizza Place   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar            Restaurant              Beer Bar   
1        Breakfast Spot             Nightclub           Coffee Shop   
2         Auto Workshop            Comic Shop      Recording Studio   
3      Airport Terminal                 Plane   Rental Car Location   
4        Sandwich Place    Italian Restaurant                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery    Seafood Restaurant           Cheese Shop   
1          Climbing Gym         Burrito Place            Restaurant   
2            Restaurant               Butcher         Burrito Place   
3         Boat or Ferry       Harbor / Marina      Sculpture Garden   
4   Japanese Restaurant           Salad Place       Bubble Tea Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café           Fish Market                  Hotel  
1    Italian Restaurant               Stadium           Intersection  
2               Brewery            Skate Park             Smoke Shop  
3              Boutique        Airport Lounge           Airport Gate  
4        Ice Cream Shop          Burger Joint                    Bar

Creating neighborhood clusters.

In [26]:
# set number of clusters
kclusters = 5

toro_grouped_clustering = toro_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toro_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [27]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toro_merged = toronto_data

# merge toro_grouped with toro_data to add latitude/longitude for each neighborhood
toro_merged = toro_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toro_merged.head()

Postalcode          Borough                    Neighborhood   Latitude  \
37        M4E     East Toronto                     The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
43        M4M     East Toronto                 Studio District  43.659526   
44        M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37 -79.293031               0                 Trail     Health Food Store   
41 -79.352188               1      Greek Restaurant    Italian Restaurant   
42 -79.315572               1  Fast Food Restaurant                   Gym   
43 -79.340923               1                  Café           Coffee Shop   
44 -79.388790               3                  Park              Bus Line   

   3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
37                   Pub               Dog Run        Department Store   
41           Coffee Shop        Ice Cream Shop  Furniture / Home Store   
42             Pet Store               Brewery          Sandwich Place   
43                Bakery             Gastropub     American Restaurant   
44           Swim School           Yoga Studio        Department Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37          Dessert Shop                 Diner        Discount Store   
41            Restaurant     Indian Restaurant                   Spa   
42         Burrito Place            Restaurant                   Pub   
43               Brewery                   Bar        Ice Cream Shop   
44           Event Space  Ethiopian Restaurant     Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
37          Distribution Center            Yoga Studio  
41                    Juice Bar                Brewery  
42                  Pizza Place                   Park  
43                  Fish Market     Italian Restaurant  
44  Eastern European Restaurant             Donut Shop

In [40]:
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               1                                        Berczy Park   
1               1       Brockton, Parkdale Village, Exhibition Place   
2               1              Business reply mail Processing Centre   
3               1  CN Tower, King and Spadina, Railway Lands, Har...   
4               1                                 Central Bay Street   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop          Cocktail Bar            Restaurant   
1                  Café        Breakfast Spot             Nightclub   
2           Pizza Place         Auto Workshop            Comic Shop   
3       Airport Service      Airport Terminal                 Plane   
4           Coffee Shop        Sandwich Place    Italian Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0              Beer Bar                Bakery    Seafood Restaurant   
1           Coffee Shop          Climbing Gym         Burrito Place   
2      Recording Studio            Restaurant               Butcher   
3   Rental Car Location         Boat or Ferry       Harbor / Marina   
4                  Café   Japanese Restaurant           Salad Place   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Cheese Shop                  Café           Fish Market   
1            Restaurant    Italian Restaurant               Stadium   
2         Burrito Place               Brewery            Skate Park   
3      Sculpture Garden              Boutique        Airport Lounge   
4       Bubble Tea Shop        Ice Cream Shop          Burger Joint   

  10th Most Common Venue  
0                  Hotel  
1           Intersection  
2             Smoke Shop  
3           Airport Gate  
4                    Bar

In [32]:
address = 'Toronto, CA'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


Create the required map with folium.

In [31]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toro_merged['Latitude'], toro_merged['Longitude'], toro_merged['Neighborhood'], toro_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters